File: client/rga.js
/* rga.js
 *
 * A direct, readable port of the Rust RGA implementation:
 * - s4vector.rs (S4Vector + precedes)
 * - remote_op.rs (RemoteOp enum)
 * - rga.rs (insert/delete/update local + remote)
 *
 * Notes:
 * - We key nodes by S4Vector string "ssn|sid|sum|seq"
 * - "obj === null" means tombstone (deleted)
 */

(function (global) {
  "use strict";

  function s4Key(s4v) {
    return `${s4v.ssn}|${s4v.sid}|${s4v.sum}|${s4v.seq}`;
  }

  // Same ordering as Rust: (ssn, sum, sid)
  function s4Precedes(a, b) {
    if (a.ssn !== b.ssn) return a.ssn < b.ssn;
    if (a.sum !== b.sum) return a.sum < b.sum;
    return a.sid < b.sid;
  }

  // Shallow clone helpers to avoid accidental aliasing
  function cloneS4(s4v) {
    return { ssn: s4v.ssn, sid: s4v.sid, sum: s4v.sum, seq: s4v.seq };
  }

  function cloneVc(vc) {
    return vc.slice();
  }

  class Rga {
    /**
     * @param {number} siteId - site_id assigned by the server
     * @param {number} numSites - total sites (vector clock length)
     */
    constructor(siteId, numSites) {
      this.session = 1;
      this.siteId = siteId;
      this.vectorClock = new Array(numSites).fill(0);

      this.head = null;          // first node in linked list
      this.nodes = new Map();    // Map<s4Key, node>
      this.cemetery = [];        // list of deleted node ids (not required for basic editing)
    }

    /** Clears the structure but keeps siteId/numSites. */
    reset() {
      this.head = null;
      this.nodes.clear();
      this.cemetery = [];
      this.vectorClock.fill(0);
    }

    /** Returns the current visible text (tombstones omitted). */
    getText() {
      const out = [];
      let cur = this.head;
      while (cur) {
        if (cur.obj !== null) out.push(cur.obj);
        cur = cur.next;
      }
      return out.join("");
    }

    /** Returns array of nodes that are NOT tombstones, in visible order. */
    visibleNodes() {
      const out = [];
      let cur = this.head;
      while (cur) {
        if (cur.obj !== null) out.push(cur);
        cur = cur.next;
      }
      return out;
    }

    /** Finds the node at a given visible index (0-based). Returns null if out of range. */
    nodeAtVisibleIndex(index) {
      if (index < 0) return null;
      let cur = this.head;
      let i = 0;
      while (cur) {
        if (cur.obj !== null) {
          if (i === index) return cur;
          i++;
        }
        cur = cur.next;
      }
      return null;
    }

    /** Generates a new S4Vector for a LOCAL op (matches Rust generate_s4vector). */
    generateS4Vector() {
      this.vectorClock[this.siteId] += 1;

      let sum = 0;
      for (const v of this.vectorClock) sum += v;

      return {
        ssn: this.session,
        sid: this.siteId,
        sum: sum,
        seq: this.vectorClock[this.siteId],
      };
    }

    /** Updates our vector clock by taking elementwise max with a remote vector_clock. */
    mergeVectorClock(remoteVc) {
      const n = Math.min(this.vectorClock.length, remoteVc.length);
      for (let i = 0; i < n; i++) {
        this.vectorClock[i] = Math.max(this.vectorClock[i], remoteVc[i]);
      }
    }

    /** Inserts a node into the linked list after "ref". */
    linkAfter(ref, node) {
      node.next = ref.next;
      ref.next = node;
    }

    /** Remote insert (matches Rust insert_remote). */
    insertRemote(leftId, value, s4v) {
      const node = {
        obj: value,     // string length 1
        s_k: cloneS4(s4v),
        s_p: cloneS4(s4v),
        next: null,
      };
      this.nodes.set(s4Key(node.s_k), node);

      // Case 1: Insert at head (left_id == null)
      if (leftId == null) {
        if (this.head == null) {
          this.head = node;
          return;
        }

        // If current head precedes new s4v, put new before head.
        // (This matches Rust: if head.s_k.precedes(&s4v) => insert at head)
        if (s4Precedes(this.head.s_k, s4v)) {
          node.next = this.head;
          this.head = node;
          return;
        }

        // Otherwise scan starting at head to find insertion point.
        let ref = this.head;
        while (ref.next && s4Precedes(s4v, ref.next.s_k)) {
          ref = ref.next;
        }
        this.linkAfter(ref, node);
        return;
      }

      // Case 2: Insert after a specific left node
      const leftNode = this.nodes.get(s4Key(leftId));
      if (!leftNode) {
        // If you hit this: your client/server do not agree on IDs for "left_id".
        throw new Error("insertRemote: left_id not found in local RGA");
      }

      let ref = leftNode;
      while (ref.next && s4Precedes(s4v, ref.next.s_k)) {
        ref = ref.next;
      }
      this.linkAfter(ref, node);
    }

    /** Remote delete (matches Rust delete_remote). */
    deleteRemote(targetId, s4v) {
      const target = this.nodes.get(s4Key(targetId));
      if (!target) throw new Error("deleteRemote: target_id not found in local RGA");

      if (target.obj !== null) this.cemetery.push(cloneS4(target.s_k));
      target.obj = null;
      target.s_p = cloneS4(s4v);
    }

    /** Remote update (matches Rust update_remote). */
    updateRemote(targetId, value, s4v) {
      const target = this.nodes.get(s4Key(targetId));
      if (!target) throw new Error("updateRemote: target_id not found in local RGA");
      if (target.obj === null) return; // tombstone: ignore updates

      // Only update if existing s_p precedes new s4v
      if (s4Precedes(target.s_p, s4v)) {
        target.obj = value;
        target.s_p = cloneS4(s4v);
      }
    }

    /**
     * Applies a remote RemoteOp envelope from the wire:
     *   { Insert: { left_id, value, s4v, vector_clock } }
     *   { Delete: { target_id, s4v, vector_clock } }
     *   { Update: { target_id, value, s4v, vector_clock } }
     */
    applyRemote(opEnvelope) {
      const kind = Object.keys(opEnvelope)[0];
      const p = opEnvelope[kind];

      // Merge vector clock first (matches Rust receive_remote_op)
      this.mergeVectorClock(p.vector_clock);

      if (kind === "Insert") {
        this.insertRemote(p.left_id, p.value, p.s4v);
      } else if (kind === "Delete") {
        this.deleteRemote(p.target_id, p.s4v);
      } else if (kind === "Update") {
        this.updateRemote(p.target_id, p.value, p.s4v);
      } else {
        throw new Error("Unknown RemoteOp kind: " + kind);
      }
    }

    /** Local insert at visible index. Returns a RemoteOp envelope to send to server. */
    insertLocal(index, valueChar) {
      const s4v = this.generateS4Vector();

      let leftId = null;
      if (index > 0) {
        const leftNode = this.nodeAtVisibleIndex(index - 1);
        if (!leftNode) throw new Error("insertLocal: index out of range");
        leftId = cloneS4(leftNode.s_k);
      }

      // Apply locally using the same remote algorithm (matches Rust insert_local)
      this.insertRemote(leftId, valueChar, s4v);

      return {
        Insert: {
          left_id: leftId,
          value: valueChar,
          s4v: cloneS4(s4v),
          vector_clock: cloneVc(this.vectorClock),
        },
      };
    }

    /** Local delete at visible index. Returns a RemoteOp envelope to send to server. */
    deleteLocal(index) {
      const target = this.nodeAtVisibleIndex(index);
      if (!target) throw new Error("deleteLocal: index out of range");

      const s4v = this.generateS4Vector();

      if (target.obj !== null) this.cemetery.push(cloneS4(target.s_k));
      target.obj = null;
      target.s_p = cloneS4(s4v);

      return {
        Delete: {
          target_id: cloneS4(target.s_k),
          s4v: cloneS4(s4v),
          vector_clock: cloneVc(this.vectorClock),
        },
      };
    }

    /** Local update (replace char) at visible index. Returns a RemoteOp envelope to send to server. */
    updateLocal(index, valueChar) {
      const target = this.nodeAtVisibleIndex(index);
      if (!target) throw new Error("updateLocal: index out of range");
      if (target.obj === null) return null;

      const s4v = this.generateS4Vector();
      target.obj = valueChar;
      target.s_p = cloneS4(s4v);

      return {
        Update: {
          target_id: cloneS4(target.s_k),
          value: valueChar,
          s4v: cloneS4(s4v),
          vector_clock: cloneVc(this.vectorClock),
        },
      };
    }

    /**
     * Bootstraps from plain text by simulating the server's initial inserts
     * (server site_id = 0, session = 1, seq=sum=1..N).
     *
     * IMPORTANT: This only works if the server's base content IDs match that scheme.
     */
    bootstrapFromPlainText(text) {
      this.reset();

      const chars = Array.from(text); // unicode code points
      let left = null;

      for (let i = 0; i < chars.length; i++) {
        const vc = new Array(this.vectorClock.length).fill(0);
        vc[0] = i + 1;

        const s4v = { ssn: 1, sid: 0, sum: i + 1, seq: i + 1 };

        this.applyRemote({
          Insert: {
            left_id: left,
            value: chars[i],
            s4v: s4v,
            vector_clock: vc,
          },
        });

        left = cloneS4(s4v);
      }
    }
  }

  global.Rga = Rga;
})(window);

File: crates/rga/src/rga.rs
use crate::node::Node;
use crate::remote_op::RemoteOp;
use crate::s4vector::S4Vector;
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

/// Main RGA structure
///
/// Uses:
/// - Linked list (via Node.link) for maintaining document order
/// - Hash map (SVI scheme) for O(1) lookup by S4Vector
/// - Rc<RefCell<>> for shared mutable access (justified: need multiple references
///   from both linked list and hash map, with interior mutability for updates)
#[derive(Debug)]
pub struct Rga<T: Clone> {
    /// Head of the linked list
    head: Option<Rc<RefCell<Node<T>>>>,

    /// Hash map for S4Vector Index (SVI) scheme - enables O(1) lookup
    /// Maps s_k -> Node reference
    hash_map: HashMap<S4Vector, Rc<RefCell<Node<T>>>>,

    /// Local site configuration
    site_id: u32,
    session: u32,
    vector_clock: Vec<u32>,

    /// Cemetery for tombstone management (Section 5.6)
    cemetery: Vec<S4Vector>,
}

impl<T: Clone> Rga<T> {
    /// Create a new RGA for the given site
    pub fn new(site_id: u32, num_sites: usize) -> Self {
        Rga {
            head: None,
            hash_map: HashMap::new(),
            site_id,
            session: 1,
            vector_clock: vec![0; num_sites],
            cemetery: Vec::new(),
        }
    }

    /// Generate S4Vector for current operation
    fn generate_s4vector(&mut self) -> S4Vector {
        self.vector_clock[self.site_id as usize] += 1;
        let sum: u32 = self.vector_clock.iter().sum();
        let seq = self.vector_clock[self.site_id as usize];

        S4Vector::new(self.session, self.site_id, sum, seq)
    }

    /// Find node by index (for local operations)
    /// Implements Algorithm 4: findlist
    /// Skips tombstones to match visible document order
    fn find_by_index(&self, index: usize) -> Option<Rc<RefCell<Node<T>>>> {
        let mut current = self.head.clone();
        let mut count = 0;

        while let Some(node_rc) = current {
            let node = node_rc.borrow();

            // Skip tombstones (they don't count in visible indices)
            if !node.is_tombstone() {
                if count == index {
                    drop(node);
                    return Some(node_rc.clone());
                }
                count += 1;
            }

            current = node.link.clone();
        }

        None
    }

    /// Find node by S4Vector (for remote operations)
    /// Uses SVI scheme for O(1) lookup
    fn find_by_s4vector(&self, s4v: &S4Vector) -> Option<Rc<RefCell<Node<T>>>> {
        self.hash_map.get(s4v).cloned()
    }

    /// Local Insert operation (Algorithm 4)
    /// Returns RemoteOp for broadcasting
    pub fn insert_local(&mut self, index: usize, value: T) -> Option<RemoteOp<T>> {
        let s4v = self.generate_s4vector();

        // Find left cobject
        let left_id = if index == 0 {
            None
        } else {
            self.find_by_index(index - 1).map(|n| n.borrow().s_k)
        };

        self.remote_insert(left_id, value.clone(), s4v);

        Some(RemoteOp::Insert {
            left_id,
            value,
            s4v,
            vector_clock: self.vector_clock.clone(),
        })
    }

    /// Local Delete operation (Algorithm 4)
    pub fn delete_local(&mut self, index: usize) -> Option<RemoteOp<T>> {
        let target = self.find_by_index(index)?;
        let target_id = target.borrow().s_k;
        let s4v = self.generate_s4vector();

        // Make tombstone
        target.borrow_mut().obj = None;
        target.borrow_mut().s_p = s4v;

        // Enroll in cemetery for later purging
        self.cemetery.push(target_id);

        Some(RemoteOp::Delete {
            target_id,
            s4v,
            vector_clock: self.vector_clock.clone(),
        })
    }

    /// Local Update operation (Algorithm 4)
    pub fn update_local(&mut self, index: usize, value: T) -> Option<RemoteOp<T>> {
        let target = self.find_by_index(index)?;
        let target_id = target.borrow().s_k;
        let s4v = self.generate_s4vector();

        // Update the node
        target.borrow_mut().obj = Some(value.clone());
        target.borrow_mut().s_p = s4v;

        Some(RemoteOp::Update {
            target_id,
            value,
            s4v,
            vector_clock: self.vector_clock.clone(),
        })
    }

    /// Read the current visible document state
    pub fn read(&self) -> Vec<T> {
        let mut result = Vec::new();
        let mut current = self.head.clone();

        while let Some(node_rc) = current {
            let node = node_rc.borrow();

            // Only include non-tombstone nodes
            if let Some(ref obj) = node.obj {
                result.push(obj.clone());
            }

            current = node.link.clone();
        }

        result
    }

    /// Apply remote operation (dispatches to specific handlers)
    /// Implements Algorithm 1 lines 16-17: update vector clock then execute
    pub fn apply_remote(&mut self, op: RemoteOp<T>) {
        // Algorithm 1 line 16: ∀k: v_i[k] := max(v_i[k], v_O[k])
        let op_vc = match &op {
            RemoteOp::Insert { vector_clock, .. } => vector_clock,
            RemoteOp::Delete { vector_clock, .. } => vector_clock,
            RemoteOp::Update { vector_clock, .. } => vector_clock,
        };

        for (i, &op_count) in op_vc.iter().enumerate() {
            if i < self.vector_clock.len() {
                self.vector_clock[i] = self.vector_clock[i].max(op_count);
            }
        }

        // Algorithm 1 line 17: RADT.remoteAlgorithm(O)
        match op {
            RemoteOp::Insert {
                left_id,
                value,
                s4v,
                ..
            } => {
                self.remote_insert(left_id, value, s4v);
            }
            RemoteOp::Delete { target_id, s4v, .. } => {
                self.remote_delete(target_id, s4v);
            }
            RemoteOp::Update {
                target_id,
                value,
                s4v,
                ..
            } => {
                self.remote_update(target_id, value, s4v);
            }
        }
    }

    /// Remote Insert operation (Algorithm 8)
    /// Implements Operation Commutativity (OC) and Precedence Transitivity (PT)
    fn remote_insert(&mut self, left_id: Option<S4Vector>, value: T, s4v: S4Vector) {
        let new_node = Rc::new(RefCell::new(Node::new(value, s4v)));

        // (i) Find left cobject via hash map - O(1)
        if let Some(left_s4v) = left_id {
            let left_node = match self.find_by_s4vector(&left_s4v) {
                Some(n) => n,
                None => {
                    // Cobject not found - should not happen with proper causality
                    eprintln!("Warning: Left cobject not found for Insert");
                    return;
                }
            };

            // (iii) Scan for correct position based on PT
            // From Algorithm 8, line 20: while(ref.link != nil and ins.s_k ≺ ref.link.s_k)
            // Continue scanning while INSERT precedes NEXT node (ins.s_k ≺ next.s_k)
            // Stop when NEXT precedes INSERT (next.s_k ≺ ins.s_k) or they're equal
            // This implements PT: succeeding inserts go closer to left cobject
            let mut ref_node = left_node.clone();
            loop {
                let next = ref_node.borrow().link.clone();

                match next {
                    Some(ref next_rc) => {
                        let next_node = next_rc.borrow();
                        // Algorithm 8 line 20: while(ins.s_k ≺ ref.link.s_k)
                        // Continue if INSERT precedes NEXT (new node should go before next)
                        // Stop if NEXT precedes INSERT (new node should go after ref, before next)
                        if s4v.precedes(&next_node.s_k) {
                            // Insert precedes next, continue scanning
                            drop(next_node);
                            ref_node = next_rc.clone();
                        } else {
                            // Next precedes or equals insert, stop here
                            drop(next_node);
                            break;
                        }
                    }
                    None => break,
                }
            }

            // (iv) Link new node into list (before the first preceding node)
            let next = ref_node.borrow().link.clone();
            new_node.borrow_mut().link = next;
            ref_node.borrow_mut().link = Some(new_node.clone());
        } else {
            // Insert at head (left_id = nil)
            // From Algorithm 8, lines 14-18
            if let Some(ref head_rc) = self.head {
                // Line 15: if(head = nil or head.s_k ≺ ins.s_k)
                if head_rc.borrow().s_k.precedes(&s4v) {
                    // Current head precedes new node, insert new node at head
                    // Lines 16-17
                    new_node.borrow_mut().link = self.head.clone();
                    self.head = Some(new_node.clone());
                } else {
                    // New node precedes head, scan forward from head
                    // Lines 19-22: same scanning logic as above
                    let mut ref_node = head_rc.clone();
                    loop {
                        let next = ref_node.borrow().link.clone();
                        match next {
                            Some(ref next_rc) => {
                                // ✅ FIXED: Continue while ins.s_k ≺ next.s_k
                                if s4v.precedes(&next_rc.borrow().s_k) {
                                    ref_node = next_rc.clone();
                                } else {
                                    break;
                                }
                            }
                            None => break,
                        }
                    }
                    // Lines 21-22
                    let next = ref_node.borrow().link.clone();
                    new_node.borrow_mut().link = next;
                    ref_node.borrow_mut().link = Some(new_node.clone());
                }
            } else {
                // Empty list - new node becomes head
                self.head = Some(new_node.clone());
            }
        }

        // (ii) Add to hash map (SVI scheme)
        self.hash_map.insert(s4v, new_node);
    }

    /// Remote Delete operation (Algorithm 9)
    /// Delete always wins (Delete 99K Update) regardless of s4vector order
    fn remote_delete(&mut self, target_id: S4Vector, s4v: S4Vector) {
        if let Some(target) = self.find_by_s4vector(&target_id) {
            let mut target_mut = target.borrow_mut();

            // Mark as tombstone (preserves cobject for future operations)
            if !target_mut.is_tombstone() {
                self.cemetery.push(target_id);
            }

            target_mut.obj = None;
            target_mut.s_p = s4v;
        } else {
            eprintln!("Warning: Target not found for Delete");
        }
    }

    /// Remote Update operation (Algorithm 10)
    /// Update only succeeds if s4v succeeds current s_p AND target is not tombstone
    /// This implements: Update 99K Delete (Updates don't resurrect tombstones)
    fn remote_update(&mut self, target_id: S4Vector, value: T, s4v: S4Vector) {
        if let Some(target) = self.find_by_s4vector(&target_id) {
            let mut target_mut = target.borrow_mut();

            // Don't update tombstones (Update 99K Delete precedence)
            if target_mut.is_tombstone() {
                return;
            }

            // Only update if new s4v succeeds current s_p (PT)
            if target_mut.s_p.precedes(&s4v) {
                target_mut.obj = Some(value);
                target_mut.s_p = s4v;
            }
        } else {
            eprintln!("Warning: Target not found for Update");
        }
    }

    /// Get current document length (excluding tombstones)
    pub fn len(&self) -> usize {
        self.read().len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

#[cfg(test)]
mod tests {
    use crate::Rga;

    #[test]
    fn test_basic_insert_and_read() {
        let mut rga = Rga::<char>::new(0, 1);

        rga.insert_local(0, 'a');
        rga.insert_local(1, 'b');
        rga.insert_local(2, 'c');

        let content = rga.read();
        assert_eq!(content, vec!['a', 'b', 'c']);
    }

    #[test]
    fn test_delete() {
        let mut rga = Rga::<char>::new(0, 1);

        rga.insert_local(0, 'a');
        rga.insert_local(1, 'b');
        rga.insert_local(2, 'c');
        rga.delete_local(1);

        let content = rga.read();
        assert_eq!(content, vec!['a', 'c']);
    }

    #[test]
    fn test_update() {
        let mut rga = Rga::<char>::new(0, 1);

        rga.insert_local(0, 'a');
        rga.insert_local(1, 'b');
        rga.update_local(1, 'x');

        let content = rga.read();
        assert_eq!(content, vec!['a', 'x']);
    }

    #[test]
    fn test_concurrent_inserts_dopt_puzzle() {
        // Example 1 from the paper: dOPT puzzle
        // Three sites concurrently insert at the same position

        let mut site0 = Rga::<char>::new(0, 3);
        let mut site1 = Rga::<char>::new(1, 3);
        let mut site2 = Rga::<char>::new(2, 3);

        // Initial state: [a, b]
        let op_a = site0.insert_local(0, 'a').unwrap();
        let op_b = site0.insert_local(1, 'b').unwrap();

        site1.apply_remote(op_a.clone());
        site1.apply_remote(op_b.clone());
        site2.apply_remote(op_a.clone());
        site2.apply_remote(op_b.clone());

        // Concurrent inserts after 'a' (index 1)
        let op1 = site0.insert_local(1, '1').unwrap(); // Site 0 inserts '1'
        let op2 = site1.insert_local(1, '2').unwrap(); // Site 1 inserts '2'
        let op3 = site2.insert_local(1, '3').unwrap(); // Site 2 inserts '3'

        // Apply in different orders at each site
        // Site 0: local op1, then remote op3, op2
        site0.apply_remote(op3.clone());
        site0.apply_remote(op2.clone());

        // Site 1: local op2, then remote op3, op1
        site1.apply_remote(op3.clone());
        site1.apply_remote(op1.clone());

        // Site 2: local op3, then remote op2, op1
        site2.apply_remote(op2.clone());
        site2.apply_remote(op1.clone());

        // All sites should converge to same order
        let result0 = site0.read();
        let result1 = site1.read();
        let result2 = site2.read();

        assert_eq!(result0, result1);
        assert_eq!(result1, result2);

        // Based on S4Vector ordering: site 0 < site 1 < site 2
        // Higher priority (later in order) goes closer to left cobject
        println!("Converged result: {:?}", result0);
    }

    #[test]
    fn test_delete_wins_over_update() {
        let mut site0 = Rga::<char>::new(0, 2);
        let mut site1 = Rga::<char>::new(1, 2);

        // Insert 'a'
        let op_insert = site0.insert_local(0, 'a').unwrap();
        site1.apply_remote(op_insert);

        // Concurrent operations: site 0 deletes, site 1 updates
        let op_delete = site0.delete_local(0).unwrap();
        let op_update = site1.update_local(0, 'x').unwrap();

        // Apply in different orders
        site0.apply_remote(op_update.clone());
        site1.apply_remote(op_delete.clone());

        // Both should have tombstone (delete wins)
        assert_eq!(site0.read(), vec![]);
        assert_eq!(site1.read(), vec![]);
    }
}
